src/hotspot/share/gc/shenandoah/heuristics/shenandoahGlobalHeuristics.cpp

```diff
@@ -53,93 +53,134 @@ void ShenandoahGlobalHeuristics::choose_global_collection_set(ShenandoahCollecti
+          max_old_cset = orig_max_old_cset;
+          committed_from_shared_reserves -= proposed_old_region_consumption * region_size_bytes;
+        }
       }
     } else {
       assert(r->is_young() && (r->age() < tenuring_threshold), "DeMorgan's law (assuming r->is_affiliated)");
-      size_t new_cset = young_cur_cset + r->get_live_data_bytes();
-      size_t region_garbage = r->garbage();
-      size_t new_garbage = cur_young_garbage + region_garbage;
-      bool add_regardless = (region_garbage > ignore_threshold) && (new_garbage < min_garbage);
-
-      if (add_regardless || (r->garbage() > garbage_threshold)) {
-        while ((new_cset > max_young_cset) && (unaffiliated_young_regions > 0)) {
-          unaffiliated_young_regions--;
+      if (add_regardless || (region_garbage > garbage_threshold)) {
+        size_t live_bytes = r->get_live_data_bytes();
+        size_t new_cset = young_cur_cset + live_bytes;
+        // May need multiple reserve regions to evacuate a single region, depending on live data bytes and ShenandoahEvacWaste
+        size_t orig_max_young_cset = max_young_cset;
+        size_t proposed_young_region_consumption = 0;
+        while ((new_cset > max_young_cset) && (committed_from_shared_reserves < shared_reserves)) {
+          committed_from_shared_reserves += region_size_bytes;
+          proposed_young_region_consumption++;
           max_young_cset += region_size_bytes / ShenandoahEvacWaste;
         }
-      }
-      if ((new_cset <= max_young_cset) && (add_regardless || (region_garbage > garbage_threshold))) {
-        add_region = true;
-        young_cur_cset = new_cset;
-        cur_young_garbage = new_garbage;
+        // We already know: add_regardless || region_garbage > garbage_threshold
+        if (new_cset <= max_young_cset) {
+          add_region = true;
+          young_cur_cset = new_cset;
+          cur_garbage = new_garbage;
+          young_evac_bytes += live_bytes;
+        } else {
+          // We failed to sufficiently expand young, so unwind proposed expansion
+          max_young_cset = orig_max_young_cset;
+          committed_from_shared_reserves -= proposed_young_region_consumption * region_size_bytes;
+        }
       }
     }
     if (add_region) {
       cset->add_region(r);
     }
   }
 
-  if (regions_transferred_to_old > 0) {
-    heap->generation_sizer()->force_transfer_to_old(regions_transferred_to_old);
-    heap->young_generation()->set_evacuation_reserve(young_evac_reserve - regions_transferred_to_old * region_size_bytes);
-    heap->old_generation()->set_evacuation_reserve(old_evac_reserve + regions_transferred_to_old * region_size_bytes);
-  }
+  heap->young_generation()->set_evacuation_reserve((size_t) (young_evac_bytes * ShenandoahEvacWaste));
+  heap->old_generation()->set_evacuation_reserve((size_t) (old_evac_bytes * ShenandoahOldEvacWaste));
+  heap->old_generation()->set_promoted_reserve((size_t) (promo_bytes * ShenandoahPromoEvacWaste));
```

Comment:
Note that the census from the most recent mark provides both these bits of information, but doesn't account for other criteria (i.e. liveness denseness) that go into the exclusion of certain regions (and their objects) from the collection set. Therein lies the rub,  but armed with historical numbers of each and reality, one might be able to predict this well (may be).